DB-backed stale worker cleanup leaves crashed hook dispatch active

## Problem
Auto-mode appeared stuck before the next task, but the runtime evidence shows it completed `execute-task M001/S01/T02` and then crashed/stalled in the post-unit sidecar hook `hook/codex-review M001/S01/T02`. The next task (`T03`) could not dispatch because the hook never emitted `unit-end` and stale DB worker state was not actually cleared.

## Root Cause
The DB-backed crash-lock cleanup path still behaves like the old file-lock era. `crash-recovery.js:139-160` removes the legacy lock file and deletes `runtime_kv` only for the current process's active worker, but it does not mark the stale worker row as crashed and does not settle the stale `unit_dispatches` row.

Problematic code:

```js
// crash-recovery.js:139-160
export function clearLock(basePath) {
    try {
        const p = lockPath(basePath);
        if (existsSync(p))
            unlinkSync(p);
    }
    catch {
        // Best-effort.
    }
    if (!isDbAvailable())
        return;
    try {
        const projectRoot = normalizeRealPath(basePath);
        const worker = findActiveWorkerForCurrentProcess(projectRoot);
        if (!worker)
            return;
        deleteRuntimeKv("worker", worker.worker_id, SESSION_FILE_KV_KEY);
    }
    catch {
        // Best-effort.
    }
}
```

For a stale worker, `findActiveWorkerForCurrentProcess()` cannot find the dead worker because its PID is not the current PID. As a result, `clearLock()` is effectively a no-op for DB-backed stale workers.

This affects both recovery paths:

```js
// doctor-runtime-checks.js:48-50
if (shouldFix("stale_crash_lock")) {
    clearLock(basePath);
    fixesApplied.push("cleared stale auto-mode worker state");
}
```

```js
// auto.js:1319-1325
if (freshStartAssessment.lock) {
    emitCrashRecoveredUnitEnd(base, freshStartAssessment.lock);
    clearLock(base);
}
```

`db/auto-workers.js:200-226` continues finding the same stale worker because the row remains `status='active'` with an expired heartbeat.

A related sidecar behavior made this visible: `auto-post-unit.js:986-994` enqueues post-unit hooks, and `auto/loop.js:382-400` drains sidecar work before normal dispatch. If `hook/codex-review` crashes, the next normal task remains blocked until the stale hook/worker state is recovered.

## Expected Behavior
When doctor or auto-start recovery detects a stale DB-backed auto worker, cleanup should durably settle the stale worker and latest dispatch, not only remove legacy file-lock state.

Concrete fix suggestion:

1. Add a DB-aware stale worker cleanup function, e.g. `clearStaleWorkerLock(basePath)` or `clearLock(basePath, { staleLock })`.
2. When `readCrashLock(basePath)` / `findStaleWorkerForProject()` identifies a stale worker:
   - mark the worker row as `crashed` using `markWorkerCrashed(worker.worker_id)`;
   - settle the worker's latest `unit_dispatches` row as `failed`, `stuck`, or `canceled` with reason `crash-recovered`;
   - delete `runtime_kv("worker", workerId, "session_file")` for that stale worker;
   - ensure subsequent `readCrashLock(basePath)` returns `null` for the same stale row.
3. Use this DB-aware cleanup from:
   - `doctor-runtime-checks.js:48-50`
   - `auto.js:1319-1325`

Suggested regression test:

- Insert a `workers` row with `status='active'`, stale heartbeat, and dead PID.
- Insert a latest `unit_dispatches` row for that worker with `status='running'`, `unit_type='hook/codex-review'`, and `unit_id='M001/S01/T02'`.
- Run doctor with `fix=true`.
- Assert worker status is no longer active, dispatch status is no longer `claimed`/`running`, and `readCrashLock(basePath)` returns `null`.

## Environment
- GSD version: 2.79.0
- Model: gpt-5.5
- Failed/stale unit: `hook/codex-review M001/S01/T02`
- User-visible symptom: auto-mode appeared stuck before `T03`

## Reproduction Context
A milestone using worktree isolation completed `execute-task M001/S01/T02`. Post-unit hooks were enabled globally with a `codex-review` hook after `execute-task`. Auto-mode enqueued and started `hook/codex-review M001/S01/T02`, then the worker process died before the hook emitted `unit-end`.

## Forensic Evidence
- Journal showed `execute-task M001/S01/T02` reached `unit-end` with `status=completed` and `artifactVerified=true`.
- Journal then showed `sidecar-dequeue` and `unit-start` for `hook/codex-review M001/S01/T02` with no matching `unit-end`.
- Metrics contained completed entries for `plan-slice M001/S01`, `execute-task M001/S01/T01`, `hook/codex-review M001/S01/T01`, and `execute-task M001/S01/T02`; it did not contain a completed entry for `hook/codex-review M001/S01/T02`.
- Doctor reported `stale_crash_lock`: stale auto-mode worker was executing `hook/codex-review M001/S01/T02` and the process was no longer running.
- Doctor also reported `.gsd.migrating`; that appears separate, but `migrate-external.js:193-199` currently marks the ambiguous “both `.gsd` and `.gsd.migrating` exist” state as fixable through doctor even though recovery returns false.

---
*Auto-generated by `/gsd forensics`*

Thanks for the triage. Confirming the bug still reproduces on the latest published version.

## Exact version
- Originally observed on `gsd-pi@2.79.0`.
- Verified still present on `gsd-pi@2.80.0` (current latest on npm). `dist/resources/extensions/gsd/crash-recovery.js` line 152 still has `clearLock()` calling `findActiveWorkerForCurrentProcess(projectRoot)` and only deleting `runtime_kv` for the *current* PID's worker. No `markWorkerCrashed` call, no stale-aware path. So a worker whose PID is dead is never settled — the diagnosis in the issue body still holds verbatim.

## Journal evidence (the stuck gap)
Trimmed from `.gsd/journal/2026-05-03.jsonl` of the affected project (irrelevant fields stripped):

```jsonl
{"ts":"2026-05-03T20:52:24.382Z","eventType":"unit-end","data":{"unitType":"execute-task","unitId":"M001/S01/T02","status":"completed","artifactVerified":true}}
{"ts":"2026-05-03T20:52:25.202Z","eventType":"iteration-end","data":{"iteration":4}}
{"ts":"2026-05-03T20:52:25.212Z","eventType":"sidecar-dequeue","data":{"kind":"hook","unitType":"hook/codex-review","unitId":"M001/S01/T02"}}
{"ts":"2026-05-03T20:52:25.214Z","eventType":"iteration-start","data":{"iteration":5}}
{"ts":"2026-05-03T20:52:25.217Z","eventType":"unit-start","data":{"unitType":"hook/codex-review","unitId":"M001/S01/T02"}}
-- 23 minutes with no further events; auto-mode is stuck here --
{"ts":"2026-05-03T21:15:41.422Z","eventType":"unit-end","data":{"unitType":"hook/codex-review","unitId":"M001/S01/T02","status":"crash-recovered","artifactVerified":false}}
```

The 23-minute gap is the symptom: `unit-start` fires at 20:52:25.217Z, the worker dies, and the `unit-end` only appears at 21:15:41.422Z after the user manually intervened. Status `crash-recovered` came from a forced recovery path, not the normal cleanup.

## Doctor output at the time
From `.gsd/forensics/report-2026-05-03-21-02-50.md` generated while still stuck:

```
[ERROR] project | stale_crash_lock | Stale auto-mode worker (PID 65560, started 2026-05-03T20:23:25.254Z, was executing hook/codex-review M001/S01/T02) — process is no longer running | file: <workers table> | fixable: yes
```

`fixable: yes` is reported, but the registered fixer is `clearLock()` — which on a stale worker is the no-op described above. Running `gsd doctor --fix` reports the issue as cleared, but the next `gsd auto` re-detects the same stale worker.

## Minimal repro
1. `gsd init` a fresh project; in `.gsd/PREFERENCES.md` set `git.isolation: worktree`.
2. Add a post-unit `codex-review` hook (any hook that sleeps a few seconds after `execute-task` is enough).
3. Start auto-mode: `gsd auto`.
4. After `unit-start hook/codex-review` appears in the journal, kill the auto-mode process with `kill -9 <pid>` (or end the terminal hard on Windows).
5. Wait ~60s for the worker heartbeat to lapse.
6. Run `gsd doctor --fix`. It reports `stale_crash_lock` as fixed.
7. Inspect the DB (e.g. `sqlite3 .gsd/gsd.db`):
   - `SELECT status FROM workers WHERE worker_id=<killed_worker>;` → still `active`.
   - `SELECT status, unit_type, unit_id FROM unit_dispatches WHERE worker_id=<killed_worker> ORDER BY id DESC LIMIT 1;` → still `running` / `claimed` on `hook/codex-review M001/S01/T02`.
   - `SELECT 1 FROM runtime_kv WHERE scope='worker' AND scope_id=<killed_worker> AND key='session_file';` → row still present.
8. `gsd auto` again — the stale worker is re-detected; loop never advances past the dead hook.

Happy to attach the raw activity JSONL (24 files) and the day's journal as a gist if it helps — let me know whether you want them.

--- a/src/resources/extensions/gsd/auto.ts
+++ b/src/resources/extensions/gsd/auto.ts
@@ -58,6 +58,7 @@ import {
 import {
   writeLock,
   clearLock,
+  clearStaleWorkerLock,
   readCrashLock,
   isLockProcessAlive,
   formatCrashInfo,
@@ -2422,7 +2423,7 @@ export async function startAuto(
     // This closes the journal gap reported in #3348 where the worker wrote side
     // effects (SUMMARY.md, DB updates) but died before emitting unit-end.
     emitCrashRecoveredUnitEnd(base, freshStartAssessment.lock);
-    clearLock(base);
+    clearStaleWorkerLock(base);
   }
 
   if (!s.paused) {

--- a/src/resources/extensions/gsd/crash-recovery.ts
+++ b/src/resources/extensions/gsd/crash-recovery.ts
@@ -30,9 +30,10 @@ import { join } from "node:path";
 import {
   findStaleWorkerForProject,
   getAllAutoWorkers,
+  markWorkerCrashed,
   type AutoWorkerRow,
 } from "./db/auto-workers.js";
-import { getLatestForUnit, type DispatchStatus } from "./db/unit-dispatches.js";
+import { markLatestActiveForWorkerCanceled, type DispatchStatus } from "./db/unit-dispatches.js";
 import { getRuntimeKv, setRuntimeKv, deleteRuntimeKv } from "./db/runtime-kv.js";
 import { _getAdapter, isDbAvailable } from "./gsd-db.js";
 import { gsdRoot, normalizeRealPath } from "./paths.js";
@@ -54,6 +55,15 @@ const SESSION_FILE_KV_KEY = "session_file";
 
 function lockPath(basePath: string): string {
   return join(gsdRoot(basePath), effectiveLockFile());
+}
+
+function clearLegacyLockFile(basePath: string): void {
+  try {
+    const p = lockPath(basePath);
+    if (existsSync(p)) unlinkSync(p);
+  } catch {
+    // Best-effort.
+  }
 }
 
 function readLegacyLock(basePath: string): LockData | null {
@@ -204,18 +214,34 @@ export function writeLock(
  * stale session-file pointer.
  */
 export function clearLock(basePath: string): void {
-  try {
-    const p = lockPath(basePath);
-    if (existsSync(p)) unlinkSync(p);
-  } catch {
-    // Best-effort.
-  }
+  clearLegacyLockFile(basePath);
 
   if (!isDbAvailable()) return;
   try {
     const projectRoot = normalizeRealPath(basePath);
     const worker = findActiveWorkerForCurrentProcess(projectRoot);
     if (!worker) return;
+    deleteRuntimeKv("worker", worker.worker_id, SESSION_FILE_KV_KEY);
+  } catch {
+    // Best-effort.
+  }
+}
+
+/**
+ * Clear a stale DB-backed worker lock after readCrashLock/findStaleWorkerForProject
+ * has identified a dead worker. Unlike clearLock(), this targets the stale
+ * worker row instead of the current process's active worker.
+ */
+export function clearStaleWorkerLock(basePath: string): void {
+  clearLegacyLockFile(basePath);
+
+  if (!isDbAvailable()) return;
+  try {
+    const projectRoot = normalizeRealPath(basePath);
+    const worker = findStaleWorkerForProject(projectRoot);
+    if (!worker) return;
+    markLatestActiveForWorkerCanceled(worker.worker_id, "crash-recovered");
+    markWorkerCrashed(worker.worker_id);
     deleteRuntimeKv("worker", worker.worker_id, SESSION_FILE_KV_KEY);
   } catch {
     // Best-effort.

--- a/src/resources/extensions/gsd/doctor-runtime-checks.ts
+++ b/src/resources/extensions/gsd/doctor-runtime-checks.ts
@@ -7,7 +7,7 @@ import { milestonesDir, gsdRoot, resolveGsdRootFile } from "./paths.js";
 import { deriveState, isGhostMilestone, isReusableGhostMilestone } from "./state.js";
 import { saveFile } from "./files.js";
 import { nativeIsRepo, nativeForEachRef, nativeUpdateRef } from "./native-git-bridge.js";
-import { readCrashLock, isLockProcessAlive, clearLock } from "./crash-recovery.js";
+import { readCrashLock, isLockProcessAlive, clearStaleWorkerLock } from "./crash-recovery.js";
 import { getActiveAutoWorkers } from "./db/auto-workers.js";
 import { normalizeRealPath } from "./paths.js";
 import { ensureGitignore, isGsdGitignored } from "./gitignore.js";
@@ -56,7 +56,7 @@ export async function checkRuntimeHealth(
         });
 
         if (shouldFix("stale_crash_lock")) {
-          clearLock(basePath);
+          clearStaleWorkerLock(basePath);
           fixesApplied.push("cleared stale auto-mode worker state");
         }
       }

--- a/src/resources/extensions/gsd/tests/crash-recovery-via-db.test.ts
+++ b/src/resources/extensions/gsd/tests/crash-recovery-via-db.test.ts
@@ -19,14 +19,15 @@ import {
   insertMilestone,
   _getAdapter,
 } from "../gsd-db.ts";
-import { registerAutoWorker } from "../db/auto-workers.ts";
+import { getAutoWorker, registerAutoWorker } from "../db/auto-workers.ts";
 import { claimMilestoneLease } from "../db/milestone-leases.ts";
-import { recordDispatchClaim } from "../db/unit-dispatches.ts";
+import { getLatestForUnit, markRunning, recordDispatchClaim } from "../db/unit-dispatches.ts";
 import { setRuntimeKv, getRuntimeKv } from "../db/runtime-kv.ts";
 import {
   writeLock,
   readCrashLock,
   clearLock,
+  clearStaleWorkerLock,
   isLockProcessAlive,
 } from "../crash-recovery.ts";
 import { normalizeRealPath } from "../paths.ts";
@@ -222,4 +223,44 @@ test("clearLock removes the session_file row for the active worker", (t) => {
   clearLock(base);
   assert.equal(getRuntimeKv("worker", workerId, "session_file"), null,
     "session_file row deleted by clearLock");
+});
+
+test("clearStaleWorkerLock crashes stale worker and cancels latest active dispatch", (t) => {
+  const base = makeBase();
+  t.after(() => cleanup(base));
+  openDatabase(join(base, ".gsd", "gsd.db"));
+  insertMilestone({ id: "M001", title: "T", status: "active" });
+  const projectRoot = normalizeRealPath(base);
+  const workerId = registerAutoWorker({ projectRootRealpath: projectRoot });
+  const lease = claimMilestoneLease(workerId, "M001");
+  assert.equal(lease.ok, true);
+  if (!lease.ok) return;
+  const claim = recordDispatchClaim({
+    traceId: "t1",
+    workerId,
+    milestoneLeaseToken: lease.token,
+    milestoneId: "M001",
+    sliceId: "S01",
+    taskId: "T02",
+    unitType: "hook/codex-review",
+    unitId: "M001/S01/T02",
+  });
+  assert.equal(claim.ok, true);
+  if (!claim.ok) return;
+  markRunning(claim.dispatchId);
+  setRuntimeKv("worker", workerId, "session_file", "/tmp/pi-session-hook.jsonl");
+  setWorkerPid(workerId, 99999);
+  expireWorker(workerId);
+
+  assert.ok(readCrashLock(base), "stale worker is detected before cleanup");
+
+  clearStaleWorkerLock(base);
+
+  assert.equal(getAutoWorker(workerId)?.status, "crashed");
+  const dispatch = getLatestForUnit("M001/S01/T02");
+  assert.ok(dispatch);
+  assert.equal(dispatch!.status, "canceled");
+  assert.equal(dispatch!.exit_reason, "crash-recovered");
+  assert.equal(getRuntimeKv("worker", workerId, "session_file"), null);
+  assert.equal(readCrashLock(base), null);
 });

--- a/src/resources/extensions/gsd/tools/plan-slice.ts
+++ b/src/resources/extensions/gsd/tools/plan-slice.ts
@@ -2,7 +2,7 @@ import { existsSync, rmSync } from "node:fs";
 import { join, relative } from "node:path";
 import { clearParseCache } from "../files.js";
 import { isClosedStatus, isDeferredStatus } from "../status-guards.js";
-import { isNonEmptyString } from "../validation.js";
+import { isNonEmptyString, validateStringArray } from "../validation.js";
 import {
   transaction,
   getMilestone,

